sycl/source/detail/program_manager/program_manager.cpp

```diff
@@ -518,6 +519,98 @@ static void emitBuiltProgramInfo(const pi_program &Prog,
   }
 }
 
+std::set<RTDeviceBinaryImage *>
+ProgramManager::collectDependentDeviceImagesForVirtualFunctions(
+    const RTDeviceBinaryImage &Img, device Dev) {
+  // If virtual functions are used in a program, then we need to link several
+  // device images together to make sure that vtable pointers stored in
+  // objects are valid between different kernels (which could be in different
+  // device images).
+  std::set<RTDeviceBinaryImage *> DeviceImagesToLink;
+  // KernelA may use some set-a, which is also used by KernelB that is in turn
+  // uses set-b, meaning that this search should be recursive. The set below
+  // is used to stop that recursion, i.e. to avoid looking at sets we have
+  // already seen.
+  std::set<std::string> HandledSets;
+  std::queue<std::string> WorkList;
+  for (const pi_device_binary_property &VFProp : Img.getVirtualFunctions()) {
+    std::string StrValue = DeviceBinaryProperty(VFProp).asCString();
+    // Device image passed to this function is expected to contain SYCL kernels
+    // and therefore it may only use virtual function sets, but cannot provide
+    // them. We expect to see just a single property here
+    assert(std::string(VFProp->Name) == "uses-virtual-functions-set" &&
+           "Unexpected virtual function property");
+    for (const auto &SetName : detail::split_string(StrValue, ',')) {
+      WorkList.push(SetName);
+      HandledSets.insert(SetName);
+    }
+  }
+
+  while (!WorkList.empty()) {
+    std::string SetName = WorkList.front();
+    WorkList.pop();
+
+    // There could be more than one device image that uses the same set
+    // of virtual functions, or provides virtual funtions from the same
+    // set.
+    for (RTDeviceBinaryImage *BinImage : m_VFSet2BinImage[SetName]) {
+      // Here we can encounter both uses-virtual-functions-set and
+      // virtual-functions-set properties, but their handling is the same: we
+      // just grab all sets they reference and add them for consideration if
+      // we haven't done so already.
+      for (const pi_device_binary_property &VFProp :
+           BinImage->getVirtualFunctions()) {
+        std::string StrValue = DeviceBinaryProperty(VFProp).asCString();
+        for (const auto &SetName : detail::split_string(StrValue, ',')) {
+          if (HandledSets.insert(SetName).second)
+            WorkList.push(SetName);
+        }
+      }
+
+      // TODO: Complete this part about handling of incompatible device images.
+      // If device image uses the same virtual function set, then we only
+      // link it if it is compatible.
+      // However, if device image provides virtual function set and it is
+      // incompatible, then should link its "dummy" version to avoid link
+      // errors about unresolved external symbols.
+      if (doesDevSupportDeviceRequirements(Dev, *BinImage))
```

Comment:
Noob question: What is meant by two device images to be incompatible? And under what circumstances, can one device image use a virtual function from another, incompatible device image?
I also don't quite understand the "dummy" version: while we can avoid link errors, but how will these dummy functions behave when they are actually invoked?

Reply:
> What is meant by two device images to be incompatible? 

I'm not sure that we define this. What we define (and check for) is if device image is compatible with a device which means that device image does not use any optional kernel features (like `fp64` or `fp16`) that are not available on that device.

> And under what circumstances, can one device image use a virtual function from another, incompatible device image?

The problem with virtual functions is that we don't know for sure if a certain virtual function is going to be used (due to indirect nature of those calls). The best example here is probably the following:

```c++
struct Base {
  virtual /* virtual function property set A */ void foo() { /* no uses of optional features */ }
  virtual /* virtual function property set B */ void bar() { /* uses fp64 */ }
};

q.single_task([=](){
  new (StoragePtr) Base;
});
```

When `Base` is constructed, a `vtable` pointer is stored into the constructed object. `vtable` is a global variable that references both `foo` and `bar`.

On the one hand, both `foo` and `bar` have to be present in a final linked device image or otherwise we will see "undefined external function" link error. On the other hand, `bar` may not be compatible with a device we are running on and subsequent kernels that perform virtual function calls may not use it, or be submitted conditionally with proper checks for device capabilities.

> I also don't quite understand the "dummy" version: while we can avoid link errors, but how will these dummy functions behave when they are actually invoked?

And here comes the "dummy" version: we both have to include `bar` and may not be able to do so and therefore (since bar is its separate set and therefore in its separate device image) we create two versions of `bar`:
- one that is original and will be linked if device supports `fp64`
- one that is "dummy". Its body is empty, i.e. it doesn't use `fp64` and can be successfully compiled on a devices without support for this feature, but we still provide the function itself just to resolve a symbol (it is expected to be unused in that case)

See  design doc in #10540 for more details. Note that the whole dynamic linking is only needed for JIT. In AOT flow, we should know our target and therefore all device images will be bundled together as necessary during compilation and we will discard unsupported virtual functions earlier.

Reply:
Ahh.. that makes a lot of sense. Thanks for the detailed explanation. I guess earlier I was looking at this from the AOT perspective, but for JIT, this makes sense.